docs/diplomat-and-macros.md

```diff
@@ -0,0 +1,272 @@
+It's not going to try and declare one better than the other, but instead just note how they differ.
+If you are reading this hoping to find an answer to "what one should I use?", then that's easy -
+each tool currently supports a unique set of foreign language bindings, so the tool you should
+use is the one that supports the languages you care about!
+
+Disclaimer: This document was written by one of the UniFFI developers, who has never used
+diplomat in anger. Please feel free to open PRs if anything here misrepresents diplomat.
+
+See also: This document was discussed in [this PR](https://github.com/mozilla/uniffi-rs/pull/1146),
+which has some very interesting discussion - indeed, some of the content in this document has
+been copy-pasted from that discussion, but there's further detail there which might be
+of interest.
+
+# The type systems
+
+The key difference between these 2 tools is the "type system". While both are exposing Rust
+code (which obviously comes with its own type system), the foreign bindings need to know
+lots of details about all the types expressed by the tool.
+
+For the sake of this document, we will use the term "type universe" to define the set of
+all types known by each of the tools. Both of these tools build their own "type universe" then
+use that to generate both Rust code and foreign bindings.
+
+## UniFFI's type universe
+
+UniFFI's model is to parse an external ffi description from a `.udl` file which describes the
+entire "type universe". This type universe is then used to generate both the Rust scaffolding
+(on disk as a `.rs` file) and the foreign bindings.
+
+**What's good about this** is that the entire type system is known when generating both the rust code
+and the foreign binding, and is known without parsing any Rust code. This is important because
+things like field names and types in structs must be known on both sides of the FFI.
+
+**What's bad about this** is that the external UDL is very ugly and redundant in terms of the
+implemented rust API.
+
+## Diplomat's type universe
+
+Diplomat defines its "type universe" (ie, the external ffi) using macros.
+
+**What's good about this** is that an "ffi module" (and there may be many) defines the canonical API
+and it is defined in terms of Rust types - the redundant UDL is removed.
+The Rust scaffolding can also be generated by the macros, meaning there are no generated `.rs`
+files involved. Types can be shared among any of the ffi modules defined in the project -
+for example, [this diplomat ffi module](https://github.com/unicode-org/icu4x/blob/7d9f89fcd7df4567e17ddd8c46810b0db287436a/ffi/diplomat/src/pluralrules.rs#L50-L51)
+uses types from a [different ffi module](https://github.com/unicode-org/icu4x/blob/7d9f89fcd7df4567e17ddd8c46810b0db287436a/ffi/diplomat/src/locale.rs#L19).
+
+It also offers better control over the stability of the API, because where the FFI is defined
```

Comment:
FWIW I don't really understand what's "better" about using a macro than a separate interface-definition file in this regard. IIUC, in UniFFI the API is stable as long as the `.udl` doesn't change, while in Diplomat the API is stable as long as the contents of hte macro block don't change.

(Making this claim doesn't bother me, I just don't get it and wonder if I'm missing something).

Reply:
Aha, I see, thanks for clarifying 👍🏻